clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java

```diff
@@ -698,36 +753,50 @@ synchronized boolean hasUnresolvedSequences() {
     // the producer id needs a reset, false otherwise.
-    private boolean shouldResetProducerStateAfterResolvingSequences() {
-        for (Iterator<TopicPartition> iter = partitionsWithUnresolvedSequences.iterator(); iter.hasNext(); ) {
+    public void checkUnresolvedSequences() {
+        for (Iterator<TopicPartition> iter = partitionsWithUnresolvedSequences.keySet().iterator(); iter.hasNext(); ) {
             TopicPartition topicPartition = iter.next();
             if (!hasInflightBatches(topicPartition)) {
-                // The partition has been fully drained. At this point, the last ack'd sequence should be once less than
+                // The partition has been fully drained. At this point, the last ack'd sequence should be one less than
                 // next sequence destined for the partition. If so, the partition is fully resolved. If not, we should
                 // reset the sequence number if necessary.
                 if (isNextSequence(topicPartition, sequenceNumber(topicPartition))) {
                     // This would happen when a batch was expired, but subsequent batches succeeded.
                     iter.remove();
                 } else {
                     // We would enter this branch if all in flight batches were ultimately expired in the producer.
-                    log.info("No inflight batches remaining for {}, last ack'd sequence for partition is {}, next sequence is {}. " +
-                            "Going to reset producer state.", topicPartition,
-                            lastAckedSequence(topicPartition).orElse(NO_LAST_ACKED_SEQUENCE_NUMBER), sequenceNumber(topicPartition));
-                    return true;
+                    if (isTransactional()) {
+                        // For the transactional producer, we bump the epoch if possible, otherwise we transition to a fatal error
+                        KafkaException exception = new KafkaException("The client hasn't received acknowledgment for " +
+                                "some previously sent messages and can no longer retry them. It isn't safe to continue.");
+                        if (canBumpEpoch()) {
+                            epochBumpRequired = true;
+                            transitionToAbortableError(exception);
+                        } else {
+                            transitionToFatalError(exception);
+                        }
+                    } else {
+                        // For the idempotent producer, reset producer state
+                        log.info("No inflight batches remaining for {}, last ack'd sequence for partition is {}, next sequence is {}. " +
+                                        "Going to bump epoch and reset sequence numbers.", topicPartition,
+                                lastAckedSequence(topicPartition).orElse(NO_LAST_ACKED_SEQUENCE_NUMBER), sequenceNumber(topicPartition));
+                        bumpEpoch();
```

Comment:
One thing I'm having a little trouble thinking through is how this epoch bump affects the other partitions, which may still have valid producer state. I guess that is the reason for the `epoch + 1` check inside `Sender.completeBatch`?

Reply:
`bumpEpoch` sets every partition's sequence number to 0. Since the epoch is higher than the last batch and the sequence is 0, the server will still accept writes for all those partitions. It's not strictly necessary for the other partitions but it really simplifies the state tracking.

Reply:
I think my question is more about how this affects retries on other partitions. As long as we are able to continue retrying with the same producerId and sequence numbers, it may not be a problem. But the logic in `canRetry` suggests that we would not.